### PR TITLE
Restore Mono runtime and class library OS synchronization context changes

### PR DESCRIPTION
**Pull request by joshpeterson**
It looks like the changes from https://github.com/Unity-Technologies/mono/pull/1169 were never brought to this branch. This pull request brings those changes here.

### CODE REVIEW

File: mono/metadata/threads-types.h [hunk 1]
```diff
@@ -464,137 +464,10 @@ mono_thread_internal_describe (MonoInternalThread *internal, GString *str);
 gboolean
 mono_thread_internal_is_current (MonoInternalThread *internal);
 
-gboolean
-mono_threads_is_current_thread_in_protected_block (void);
```
Comment (TautvydasZilys) [inline mono/metadata/threads-types.h]: Was the removal of all these APIs intentional?

File: mono/metadata/icall.c [hunk 2]
```diff
@@ -9844,9 +9844,15 @@ ves_icall_System_Net_NetworkInformation_LinuxNetworkChange_CloseNLSocket (gpoint
 
 #include "metadata/icall-def.h"
 
-#undef HANDLES
-#undef HANDLES_REUSE_WRAPPER
-#undef ICALL_TYPE
-#undef ICALL
-#undef NOHANDLES
-#undef MONO_HANDLE_REGISTER_ICALL
```
Comment (TautvydasZilys) [inline mono/metadata/icall.c:9852]: This seems like an unintended change?

File: mono/metadata/threads-types.h (hunk 1 above)
Comment (joshpeterson) [reply mono/metadata/threads-types.h]: No, it looks like I messed up the merge when I cherry-pick the changes. I'll fix this.

File: mono/metadata/icall.c (hunk 2 above)
Comment (joshpeterson) [reply mono/metadata/icall.c:9852]: Yes, this is the same as above, I'll review the merge I did.

Comment (joshpeterson) [reply mono/metadata/icall.c:9852]: Ok, I've fixed the changes. This diff looks much better now.

**Review by TautvydasZilys — APPROVED** [review]

File: mcs/class/referencesource/mscorlib/system/threading/synchronizationcontext.cs [hunk 3]
```diff
@@ -297,6 +297,11 @@ private static SynchronizationContext GetThreadLocalContext()
                 context = AndroidPlatform.GetDefaultSyncContext ();
 #endif
 
+#if UNITY_AOT
```
Comment (joncham) [inline mcs/class/referencesource/mscorlib/system/threading/synchronizationcontext.cs:300]: Why only UNITY_AOT?

**Review by joncham — APPROVED** [review]

File: mcs/class/referencesource/mscorlib/system/threading/synchronizationcontext.cs (hunk 3 above)
Comment (joshpeterson) [reply mcs/class/referencesource/mscorlib/system/threading/synchronizationcontext.cs:300]: I believe that the original change was only for IL2CPP. @TautvydasZilys do you recall?

Comment (TautvydasZilys) [reply mcs/class/referencesource/mscorlib/system/threading/synchronizationcontext.cs:300]: Indeed, this was specifically for Windows Runtime support on Xbox One and UWP, both of which are IL2CPP only.

**Review by UnityAlex — APPROVED** [review]

**Review by bholmes — APPROVED** [review]

**Pull request merged**